### PR TITLE
feat(fauna): update adapter for FQL v10

### LINKED ISSUE #8394: Fauna v10 support
### Description 📓

[Fauna released a major change with v10 this week](https://docs.fauna.com/fauna/current/migration). With it is a new version of FQL that is significantly different (and simpler) from the old version. I'm wondering if it would make sense to create a new adapter rather than migrate the existing one?

### How to reproduce ☕️

none.

### Contributing 🙌🏽

Yes, I am willing to help implement this feature in a PR

### PR DESCRIPTION
**Pull request by employee451**
This PR tries to do the same as https://github.com/nextauthjs/next-auth/pull/8403. Rather than make a PR to @linzjax's branch, i thought it would be easier to open a new PR.

## ☕️ Reasoning

Fauna's new query language, FQL v10, has been out for a while. This PR updates `adapter-fauna` to use FQL v10 and the new conventions that have come from Fauna's side.

### Breakdown of changes:

- Use singular, uppercase names for collections (this is the new naming schema Fauna uses [in their examples](https://docs.fauna.com/fauna/current/reference/fsl_reference/syntax))
- Stop using `fauna-schema-migrate`, and instead use the Fauna CLI and [FSL](https://docs.fauna.com/fauna/current/reference/fsl_reference/syntax) for defining the collections and indexes
- Switch from the `faunadb` to the `fauna` package and refactor all queries from FQL v4 to FQL v10
- Update the documentation and add a "Migrating from v1" section

## 🧢 Checklist

- [x] Documentation
- [x] Tests
- [x] Ready to be merged

## 🎫 Affected issues

Fixes: https://github.com/nextauthjs/next-auth/issues/8394

## 📌 Resources

- [Security guidelines](https://github.com/nextauthjs/.github/blob/main/SECURITY.md)
- [Contributing guidelines](https://github.com/nextauthjs/.github/blob/main/CONTRIBUTING.md)
- [Code of conduct](https://github.com/nextauthjs/.github/blob/main/CODE_OF_CONDUCT.md)
- [Contributing to Open Source](https://kcd.im/pull-request)

### CODE REVIEW

**Comment by employee451** [comment]
Hi @ndom91 & @balazsorban44. It would be amazing if we could get this merged. We are currently migrating our app to `next-auth` and we want to use the new naming conventions Fauna uses for collections. The tests are all passing🌞
Thanks for your work on Auth.js!

File: packages/adapter-fauna/src/index.ts [hunk 1]
```diff
@@ -10,140 +9,68 @@
...
-      if (
-        (error as errors.FaunaError).description?.includes(
-          "Number or numeric String expected"
-        )
-      )
-        return null
+export type FaunaUser = ToFauna<AdapterUser>
+export type FaunaSession = ToFauna<AdapterSession>
+export type FaunaVerificationToken = ToFauna<VerificationToken>
+export type FaunaAccount = ToFauna<AdapterAccount> & any // TODO: Remove `& any`
```
Comment (balazsorban44) [inline packages/adapter-fauna/src/index.ts:34]: I could not drop the any here. Any ideas why?

File: packages/adapter-fauna/src/index.ts [hunk 2]
```diff
@@ -10,140 +9,68 @@
...
+sessionToken,
+userId,
+expires
+`
+
+export const verificationTokenFields = fql`
+identifier,
+expires,
+token
+`
```
Comment (balazsorban44) [inline packages/adapter-fauna/src/index.ts:54]: This should not be hardcoded. Any ideas on how to achieve this? The fields should be driven by the object keys of the argument of the given adapter method. Maybe `Object.keys(obj).join(",\n")` or something similar can help here.

**Review by balazsorban44 — CHANGES_REQUESTED** [review]
Unfortunately tests did _not_ pass. <picture data-single-emoji=":thinking:" title=":thinking:">:thinking:</picture> I tried fixing them (missing `fauna-shell`, too short `sleep`), but I get:

```sh
test/index.test.ts > createUser
     → Response is in an unkown format: [object Object]
   ❯ test/index.test.ts > getUser
     → Response is in an unkown format: [object Object]
   ❯ test/index.test.ts > getUserByEmail
     → Response is in an unkown format: [object Object]
   ❯ test/index.test.ts > createSession
     → Response is in an unkown format: [object Object]
   ❯ test/index.test.ts > getSessionAndUser
     → Response is in an unkown format: [object Object]
   ❯ test/index.test.ts > updateUser
     → Response is in an unkown format: [object Object]
   ❯ test/index.test.ts > updateSession
     → Response is in an unkown format: [object Object]
   ❯ test/index.test.ts > linkAccount
     → Response is in an unkown format: [object Object]
   ❯ test/index.test.ts > getUserByAccount
     → Response is in an unkown format: [object Object]
   ❯ test/index.test.ts > deleteSession
     → Response is in an unkown format: [object Object]
   ❯ test/index.test.ts > createVerificationToken
     → Response is in an unkown format: [object Object]
   ❯ test/index.test.ts > useVerificationToken
     → Response is in an unkown format: [object Object]
   ❯ test/index.test.ts > unlinkAccount
     → Response is in an unkown format: [object Object]
   ❯ test/index.test.ts > deleteUser
     → Response is in an unkown format: [object Object]
```

(Looks like the error message of `fauna` has a typo :grimacing:, `unkown` -> `unknown` )

Could you have a look and make the tests pass?

I also refactored the code as described in https://authjs.dev/guides/adapters/creating-a-database-adapter#official-adapter-guidelines

Most importantly, adapter methods should not assume any field names so it remains flexible.

Formatting should happen the way in and out of the database too. Auth.js assumes plain JS values.

File: packages/adapter-fauna/src/index.ts (hunk 2 above)
Comment (employee451) [reply packages/adapter-fauna/src/index.ts]: Fixed in https://github.com/nextauthjs/next-auth/pull/9849/commits/93fb4531ffe021215b4f4236ad03c70647d9e6ef

File: packages/utils/adapter.ts [hunk 3]
```diff
@@ -65,7 +65,7 @@ export async function runBasicTests(options: TestOptions) {
   })
 
   let user = options.fixtures?.user ?? {
-    id: randomUUID(),
+    id: id(),
```
Comment (employee451) [inline packages/utils/adapter.ts:68]: @balazsorban44 I made these changes in this file. We needed to specify the custom function for generating ID's (see `packages/adapter-fauna/test/index.test.ts` in this PR). But the custom function an adapter could specify was only used once in this file and not in the other cases which I changed here

Comment (employee451) [reply packages/utils/adapter.ts:68]: Should I open a new PR for this change?

File: packages/adapter-fauna/src/index.ts (hunk 1 above)
Comment (employee451) [reply packages/adapter-fauna/src/index.ts:34]: Skærmbillede 2024-02-05 kl  10 37 15

It has to do with the last line in this interface (`readonly [parameter: string]: JsonValue | undefined;`). 
Even if I change it to be `readonly [parameter: string]: string;`, it still doesn't work (even though that should be compatible with Fauna's `QueryValue` type (`undefined` is not a valid type for `QueryValue`). If I remove the last line, however, the type error disappears.

File: packages/adapter-fauna/package.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@auth/fauna-adapter",
-  "version": "1.3.2",
+  "version": "2.0.0",
```
Comment (employee451) [inline packages/adapter-fauna/package.json:3]: As discussed in this comment https://github.com/nextauthjs/next-auth/issues/8394#issuecomment-1690655534, we should make a major version bump for the package. People who previously used the adapter need to rename their Fauna collections to make 2.0 work. 
I also considered accepting a config in the adapter, specifying what the collection names are. These could then be used with string interpolation, so people can use any naming schema they want for their collections (or keep the old ones). Let me know what you think.

**Comment by employee451** [comment]
Sorry if I'm being pushy. We really want to get this merged so we can switch from Auth0 to next-auth with an email provider. I've made the changes you requested and the tests are passing properly now ✅ 🌞
@balazsorban44 @ndom91 @ThangHuuVu

File: packages/adapter-fauna/package.json (hunk 4 above)
Comment (employee451) [reply packages/adapter-fauna/package.json:3]: If you think the collection names config sounds like a good idea, I'd like to open a new PR for this after we get this merged

**Comment by ndom91** [comment]
Hey @employee451 first of all thanks for the contribution! 

Can you give me a short run down of the implications of this for existing fauna adapter users?

**Comment by employee451** [comment]
> Hey @employee451 first of all thanks for the contribution!
> 
> Can you give me a short run down of the implications of this for existing fauna adapter users?

Thanks for your reply @ndom91!
Existing Fauna adapter users will need to use a Fauna client from the new `fauna` library instead of the old `faunadb`. Additionally, they will have to rename the auth collections that the adapter uses. This is explained in the new "Migrating from v1" section I've added to the adapter docs:

```
In v2, we've renamed the collections to use uppercase naming, in accordance with Fauna best practices. If you're migrating from v1, you'll need to rename your collections to match the new naming scheme.
Additionally, we've renamed the indexes to match the new method-like index names. Please see above for the new index definitions.
```

File: packages/adapter-fauna/src/index.ts (hunk 1 above)
Comment (ndom91) [reply packages/adapter-fauna/src/index.ts:34]: @employee451 could you take another look at this todo? Would be great to get that `any` out of there :pray:

Comment (employee451) [reply packages/adapter-fauna/src/index.ts:34]: @ndom91 I have just pushed a fix for this. Please take a look

**Review by ndom91 — APPROVED** [review]
LGTM :tada:

**Comment by employee451** [comment]
> LGTM 🎉

Thanks! What needs to happen to get this merged? Do we need @balazsorban44 to approve the changes?

**Comment by ndom91** [comment]
@employee451 what do you think about this migration script in FQL? Looks like it should do the trick to me:

```fauna
Collection.byName("accounts")!.update({
  name: "Account"
  indexes: {
    byUserId: {
        terms: [{ field: "userId" }]
    },
    byProviderAndProviderAccountId: {
        terms: [{ field: "provider" }, { field: "providerAccountId" }]
    }
  }
})
Collection.byName("sessions")!.update({
  name: "Session",
  indexes: {
    bySessionToken: {
        terms: [{ field: "sessionToken" }]
    },
    byUserId: {
        terms: [{ field: "userId" }]
    },
  }
})
Collection.byName("users")!.update({
  name: "User",
  indexes: {
    byEmail: {
        terms: [{ field: "email" }]
    },
  }
})
Collection.byName("verification_tokens")!.update({
  name: "VerificationToken",
  indexes: {
    byIdentifierAndToken: {
        terms: [{ field: "identifier" }, { field: "token" }]
    },
  }
})
```

I can merge it, but we'll need someone else to cut a release. We plan on doign that this weekend anyway. No need to tag anyone else :wink: :+1:

**Comment by employee451** [comment]
> @employee451 what do you think about this migration script in FQL? Looks like it should do the trick to me:
> 
> ```
> Collection.byName("accounts")!.update({
>   name: "Account"
>   indexes: {
>     byUserId: {
>         terms: [{ field: "userId" }]
>     },
>     byProviderAndProviderAccountId: {
>         terms: [{ field: "provider" }, { field: "providerAccountId" }]
>     }
>   }
> })
> Collection.byName("sessions")!.update({
>   name: "Session",
>   indexes: {
>     bySessionToken: {
>         terms: [{ field: "sessionToken" }]
>     },
>     byUserId: {
>         terms: [{ field: "userId" }]
>     },
>   }
> })
> Collection.byName("users")!.update({
>   name: "User",
>   indexes: {
>     byEmail: {
>         terms: [{ field: "email" }]
>     },
>   }
> })
> Collection.byName("verification_tokens")!.update({
>   name: "VerificationToken",
>   indexes: {
>     byIdentifierAndToken: {
>         terms: [{ field: "identifier" }, { field: "token" }]
>     },
>   }
> })
> ```
> 
> I can merge it, but we'll need someone else to cut a release. We plan on doign that this weekend anyway. No need to tag anyone else 😉 👍

The script looks good😄

**Comment by ndom91** [comment]
Ah forgot to drop the existing indexes, here's a slightly updated one: 

```ts
Collection.byName("accounts")!.update({
  name: "Account"
  indexes: {
    byUserId: {
        terms: [{ field: "userId" }]
    },
    byProviderAndProviderAccountId: {
        terms: [{ field: "provider" }, { field: "providerAccountId" }]
    },
    account_by_provider_and_provider_account_id: null,
    accounts_by_user_id: null
  }
})
Collection.byName("sessions")!.update({
  name: "Session",
  indexes: {
    bySessionToken: {
        terms: [{ field: "sessionToken" }]
    },
    byUserId: {
        terms: [{ field: "userId" }]
    },
    session_by_session_token: null,
    sessions_by_user_id: null
  }
})
Collection.byName("users")!.update({
  name: "User",
  indexes: {
    byEmail: {
        terms: [{ field: "email" }]
    },
    user_by_email: null
  }
})
Collection.byName("verification_tokens")!.update({
  name: "VerificationToken",
  indexes: {
    byIdentifierAndToken: {
        terms: [{ field: "identifier" }, { field: "token" }]
    },
    verification_token_by_identifier_and_token: null
  }
})
```

If you also think its good, lets include it in the docs. What do you think?

**Comment by ndom91** [comment]
Awesome, if you cuold add that to this PR, i'll approve it again :+1:

**Review by ndom91 — APPROVED** [review]

**Comment by ndom91** [comment]
Okay merged it, thanks again! Like I said, we plan on getting a release out this weekend, so I'll keep you posted :+1:

**Comment by employee451** [comment]
Thank you so much! I'm excited for v5🍃

**Comment by ndom91** [comment]
@employee451 available in `@auth/fauna-adapter@2.1.0` now :tada:

**Pull request merged**